File: sdk/ios/Classes/MyTrackerSDKPlugin.m
#import "MyTrackerSDKPlugin.h"
#import <MyTrackerSDK/MyTrackerSDK.h>

static NSString *apiChannelName = @"_mytracker_api_channel";

static NSString *initMethod = @"init";
static NSString *trackEventMethod = @"trackEvent";
static NSString *trackLoginEventMethod = @"trackLoginEvent";
static NSString *trackRegistrationEventMethod = @"trackRegistrationEvent";
static NSString *trackLevelEventMethod = @"trackLevelEvent";
static NSString *trackInviteEventMethod = @"trackInviteEvent";
static NSString *flushMethod = @"flush";
static NSString *getInstanceIdMethod = @"getInstanceId";
static NSString *isDebugModeMethod = @"isDebugMode";
static NSString *setDebugModeMethod = @"setDebugMode";
static NSString *getIdMethod = @"getId";
static NSString *setAttributionDelegateMethod = @"setAttributionListener";
static NSString *handleDeeplinkMethod = @"handleDeeplink";

static NSString *getBufferingPeriodMethod = @"getBufferingPeriod";
static NSString *getForcingPeriodMethod = @"getForcingPeriod";
static NSString *getLaunchTimeoutMethod = @"getLaunchTimeout";
static NSString *isTrackingEnvironmentEnabledMethod = @"isTrackingEnvironmentEnabled";
static NSString *isTrackingLaunchEnabledMethod = @"isTrackingLaunchEnabled";
static NSString *isTrackingLocationEnabledMethod = @"isTrackingLocationEnabled";
static NSString *setBufferingPeriodMethod = @"setBufferingPeriod";
static NSString *setForcingPeriodMethod = @"setForcingPeriod";
static NSString *setLaunchTimeoutMethod = @"setLaunchTimeout";
static NSString *setProxyHostMethod = @"setProxyHost";
static NSString *setTrackingEnvironmentEnabledMethod = @"setTrackingEnvironmentEnabled";
static NSString *setTrackingLaunchEnabledMethod = @"setTrackingLaunchEnabled";
static NSString *setTrackingLocationEnabledMethod = @"setTrackingLocationEnabled";

static NSString *getAgeMethod = @"getAge";
static NSString *getGenderMethod = @"getGender";
static NSString *getLangMethod = @"getLang";
static NSString *getCustomUserIdMethod = @"getCustomUserIds";
static NSString *getEmailMethod = @"getEmails";
static NSString *getPhoneMethod = @"getPhones";
static NSString *setAgeMethod = @"setAge";
static NSString *setGenderMethod = @"setGender";
static NSString *setLangMethod = @"setLang";
static NSString *setCustomUserIdMethod = @"setCustomUserIds";
static NSString *setEmailMethod = @"setEmails";
static NSString *setPhoneMethod = @"setPhones";
static NSString *setCustomParamMethod = @"setCustomParam";
static NSString *getCustomParamMethod = @"getCustomParam";

static NSString *idParam = @"id";
static NSString *userIdParam = @"userId";
static NSString *vkConnectIdParam = @"vkConnectId";
static NSString *nameParam = @"name";
static NSString *eventParamsParam = @"eventParams";
static NSString *keyParam = @"key";
static NSString *valueParam = @"value";
static NSString *levelParam = @"level";
static NSString *uriParam = @"uri";

static NSString *onReceiveAttributionCallback = @"onReceiveAttribution";

@interface MyTrackerSDKPlugin () <MRMyTrackerAttributionDelegate>

@end

@implementation MyTrackerSDKPlugin
{
	FlutterMethodChannel *_apiChannel;
}

+ (void)registerWithRegistrar:(NSObject <FlutterPluginRegistrar> *)registrar
{
	MyTrackerSDKPlugin *myTrackerSdkPlugin = [[MyTrackerSDKPlugin alloc] initWithMessenger:[registrar messenger]];

	FlutterMethodChannel *apiChannel = myTrackerSdkPlugin->_apiChannel;
	if (!apiChannel)
	{
		return;
	}

	[registrar addMethodCallDelegate:myTrackerSdkPlugin channel:apiChannel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result
{
	id returnValue = nil;

	NSString *method = [call method];
	id arguments = call.arguments;
	if ([initMethod isEqualToString:method])
	{
		[MRMyTracker setupTracker:arguments[idParam]];
	}
	else if ([flushMethod isEqualToString:method])
	{
		[MRMyTracker flush];
	}
	else if ([getInstanceIdMethod isEqualToString:method])
	{
		returnValue = [MRMyTracker instanceId];
	}
	else if ([trackEventMethod isEqualToString:method])
	{
		[MRMyTracker trackEventWithName:arguments[nameParam] eventParams:[self getEventParams:arguments]];
	}
	else if ([trackLoginEventMethod isEqualToString:method])
	{
		[MRMyTracker trackLoginEvent:arguments[userIdParam] withVkConnectId:arguments[vkConnectIdParam] params:[self getEventParams:arguments]];
	}
	else if ([trackRegistrationEventMethod isEqualToString:method])
	{
		[MRMyTracker trackRegistrationEvent:arguments[userIdParam] withVkConnectId:arguments[vkConnectIdParam] params:[self getEventParams:arguments]];
	}
	else if ([trackLevelEventMethod isEqualToString:method])
	{
		[MRMyTracker trackLevelAchievedWithLevel:arguments[levelParam] eventParams:[self getEventParams:arguments]];
	}
	else if ([trackInviteEventMethod isEqualToString:method])
	{
		[MRMyTracker trackInviteEventWithParams:[self getEventParams:arguments]];
	}
	else if ([isDebugModeMethod isEqualToString:method])
	{
		returnValue = @(MRMyTracker.isDebugMode);
	}
	else if ([setDebugModeMethod isEqualToString:method])
	{
		[MRMyTracker setDebugMode:((NSNumber *) arguments[valueParam]).boolValue];
	}
	else if ([getIdMethod isEqualToString:method])
	{
		returnValue = MRMyTracker.trackerConfig.trackerId;
	}
	else if ([getBufferingPeriodMethod isEqualToString:method])
	{
		returnValue = @((int) MRMyTracker.trackerConfig.bufferingPeriod);
	}
	else if ([getForcingPeriodMethod isEqualToString:method])
	{
		returnValue = @((int) MRMyTracker.trackerConfig.forcingPeriod);
	}
	else if ([getLaunchTimeoutMethod isEqualToString:method])
	{
		returnValue = @((int) MRMyTracker.trackerConfig.launchTimeout);
	}
	else if ([isTrackingEnvironmentEnabledMethod isEqualToString:method])
	{
		returnValue = @(MRMyTracker.trackerConfig.trackEnvironment);
	}
	else if ([isTrackingLaunchEnabledMethod isEqualToString:method])
	{
		returnValue = @(MRMyTracker.trackerConfig.trackLaunch);
	}
	else if ([isTrackingLocationEnabledMethod isEqualToString:method])
	{
		returnValue = @((BOOL) (MRMyTracker.trackerConfig.locationTrackingMode != MRLocationTrackingModeNone));
	}
	else if ([setBufferingPeriodMethod isEqualToString:method])
	{
		MRMyTracker.trackerConfig.bufferingPeriod = ((NSNumber *) arguments[valueParam]).doubleValue;
	}
	else if ([setForcingPeriodMethod isEqualToString:method])
	{
		MRMyTracker.trackerConfig.forcingPeriod = ((NSNumber *) arguments[valueParam]).doubleValue;
	}
	else if ([setLaunchTimeoutMethod isEqualToString:method])
	{
		MRMyTracker.trackerConfig.launchTimeout = ((NSNumber *) arguments[valueParam]).doubleValue;
	}
	else if ([setProxyHostMethod isEqualToString:method])
	{
		MRMyTracker.trackerConfig.proxyHost = [self checkedCastWithClass:NSString.class value:arguments[valueParam]];
	}
	else if ([setTrackingEnvironmentEnabledMethod isEqualToString:method])
	{
		MRMyTracker.trackerConfig.trackEnvironment = ((NSNumber *) arguments[valueParam]).boolValue;
	}
	else if ([setTrackingLaunchEnabledMethod isEqualToString:method])
	{
		MRMyTracker.trackerConfig.trackLaunch = ((NSNumber *) arguments[valueParam]).boolValue;
	}
	else if ([setTrackingLocationEnabledMethod isEqualToString:method])
	{
		MRMyTracker.trackerConfig.locationTrackingMode = ((NSNumber *) arguments[valueParam]).boolValue ? MRLocationTrackingModeCached
		                                                                                                : MRLocationTrackingModeNone;
	}
	else if ([getAgeMethod isEqualToString:method])
	{
		returnValue = MRMyTracker.trackerParams.age;
	}
	else if ([getGenderMethod isEqualToString:method])
	{
		returnValue = @((int) MRMyTracker.trackerParams.gender + 1);
	}
	else if ([getLangMethod isEqualToString:method])
	{
		returnValue = MRMyTracker.trackerParams.language;
	}
	else if ([getCustomUserIdMethod isEqualToString:method])
	{
		returnValue = MRMyTracker.trackerParams.customUserIds.copy;
	}
	else if ([getEmailMethod isEqualToString:method])
	{
		returnValue = MRMyTracker.trackerParams.emails.copy;
	}
	else if ([getPhoneMethod isEqualToString:method])
	{
		returnValue = MRMyTracker.trackerParams.phones.copy;
	}
	else if ([setAgeMethod isEqualToString:method])
	{
		MRMyTracker.trackerParams.age = (NSNumber *) arguments[valueParam];
	}
	else if ([setGenderMethod isEqualToString:method])
	{
		MRMyTracker.trackerParams.gender = (MRGender) ((NSNumber *) arguments[valueParam]).intValue - 1;
	}
	else if ([setLangMethod isEqualToString:method])
	{
		MRMyTracker.trackerParams.language = arguments[valueParam];
	}
	else if ([setCustomUserIdMethod isEqualToString:method])
	{
		MRMyTracker.trackerParams.customUserIds = ((NSArray *) arguments[valueParam]).copy;
	}
	else if ([setEmailMethod isEqualToString:method])
	{
		MRMyTracker.trackerParams.emails = ((NSArray *) arguments[valueParam]).copy;
	}
	else if ([setPhoneMethod isEqualToString:method])
	{
		MRMyTracker.trackerParams.phones = ((NSArray *) arguments[valueParam]).copy;
	}
	else if ([setCustomParamMethod isEqualToString:method])
	{
		// Not available for iOS
		NSLog(@"MyTrackerSDKPlugin: %@ method is not available for iOS", setCustomParamMethod);
	}
	else if ([getCustomParamMethod isEqualToString:method])
	{
		// Not available for iOS
		NSLog(@"MyTrackerSDKPlugin: %@ method is not available for iOS", getCustomParamMethod);
	}
	else if ([setAttributionDelegateMethod isEqualToString:method])
	{
		[MRMyTracker setAttributionDelegate:self];
	}
	else if ([handleDeeplinkMethod isEqualToString:method])
	{
		NSString *urlString = arguments[uriParam];
		if (urlString == nil)
		{
			return;
		}
		NSURL *url = [NSURL URLWithString:urlString];
		if (url == nil)
		{
			return;
		}
		[MRMyTracker handleOpenURL:url options:nil];
	}
	else
	{
		returnValue = FlutterMethodNotImplemented;
	}

	result(returnValue);
}

- (instancetype)initWithMessenger:(nonnull NSObject <FlutterBinaryMessenger> *)messenger
{
	self = [super init];
	if (self)
	{
		_apiChannel = [FlutterMethodChannel methodChannelWithName:apiChannelName binaryMessenger:messenger];
	}
	return self;
}

- (id)checkedCastWithClass:(Class)class value:(id)value
{
	if (value && [value isKindOfClass:class])
	{
		return value;
	}

	return nil;
};

- (NSDictionary *)getEventParams:(id)arguments
{
	return [self checkedCastWithClass:NSDictionary.class value:arguments[eventParamsParam]];
}

- (void)didReceiveAttribution:(nonnull MRMyTrackerAttribution *)attribution 
{
	[_apiChannel invokeMethod:onReceiveAttributionCallback arguments:attribution.deeplink];
}

@end

File: sdk/android/src/main/kotlin/com/my/tracker/MyTrackerSDKPlugin.kt
package com.my.tracker

import android.app.Activity
import android.app.Application
import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.lang.ref.WeakReference

class MyTrackerSDKPlugin : FlutterPlugin, ActivityAware, MethodCallHandler {

    private var contextRef: WeakReference<Context>? = null
    private var activityRef: WeakReference<Activity>? = null

    private var myTrackerChannel: MethodChannel? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        val methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, CHANNEL_NAME)
        methodChannel.setMethodCallHandler(this)
        myTrackerChannel = methodChannel

        contextRef = WeakReference(flutterPluginBinding.applicationContext)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        myTrackerChannel?.setMethodCallHandler(null)
        myTrackerChannel = null
        contextRef = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityRef = WeakReference(binding.activity)
    }

    override fun onDetachedFromActivityForConfigChanges() = Unit

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) = Unit

    override fun onDetachedFromActivity() {
        activityRef = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        val value: Any? = when (call.method) {
            INIT_METHOD -> init(call)
            FLUSH_METHOD -> MyTracker.flush()
			GET_INSTANCE_ID -> contextRef?.get()?.let { MyTracker.getInstanceId(it) }
            TRACK_EVENT_METHOD -> MyTracker.trackEvent(call.argument(TRACK_EVENT_NAME)!!, call.argument(TRACK_EVENT_PARAMS))
            TRACK_LOGIN_METHOD -> MyTracker.trackLoginEvent(call.argument(TRACK_USERID)!!, call.argument(TRACK_VKCONNECTID), call.argument(TRACK_EVENT_PARAMS))
            TRACK_REGISTRATION_METHOD -> MyTracker.trackRegistrationEvent(call.argument(TRACK_USERID)!!, call.argument(TRACK_VKCONNECTID), call.argument(TRACK_EVENT_PARAMS))
            IS_DEBUG_MODE_METHOD -> MyTracker.isDebugMode()
            SET_DEBUG_MODE_METHOD -> MyTracker.setDebugMode(call.argument(VALUE)!!)
            GET_ID_METHOD -> MyTracker.getTrackerConfig().id
            GET_BUFFERING_PERIOD_METHOD -> MyTracker.getTrackerConfig().bufferingPeriod
            SET_BUFFERING_PERIOD_METHOD -> MyTracker.getTrackerConfig().bufferingPeriod = call.argument(VALUE)!!
            GET_FORCING_PERIOD_METHOD -> MyTracker.getTrackerConfig().forcingPeriod
            SET_FORCING_PERIOD_METHOD -> MyTracker.getTrackerConfig().forcingPeriod = call.argument(VALUE)!!
            GET_LAUNCH_TIMEOUT_METHOD -> MyTracker.getTrackerConfig().launchTimeout
            SET_LAUNCH_TIMEOUT_METHOD -> MyTracker.getTrackerConfig().launchTimeout = call.argument(VALUE)!!
            SET_PROXY_HOST_METHOD -> MyTracker.getTrackerConfig().setProxyHost(call.argument(VALUE)).let { }
            IS_TRACKING_ENVIRONMENT_ENABLED -> MyTracker.getTrackerConfig().isTrackingEnvironmentEnabled
            SET_TRACKING_ENVIRONMENT_ENABLED -> MyTracker.getTrackerConfig().isTrackingEnvironmentEnabled = call.argument(VALUE)!!
            IS_TRACKING_LAUNCH_ENABLED -> MyTracker.getTrackerConfig().isTrackingLaunchEnabled
            SET_TRACKING_LAUNCH_ENABLED -> MyTracker.getTrackerConfig().isTrackingLaunchEnabled = call.argument(VALUE)!!
            IS_TRACKING_LOCATION_ENABLED -> MyTracker.getTrackerConfig().isTrackingLocationEnabled
            SET_TRACKING_LOCATION_ENABLED -> MyTracker.getTrackerConfig().isTrackingLocationEnabled = call.argument(VALUE)!!
            SET_AGE -> MyTracker.getTrackerParams().age = call.argument(VALUE) ?: -1
            GET_AGE -> MyTracker.getTrackerParams().age.takeIf { it != -1 }
            SET_GENDER -> MyTracker.getTrackerParams().gender = call.argument<Int>(VALUE)!! - 1
            GET_GENDER -> MyTracker.getTrackerParams().gender + 1
            SET_LANG -> MyTracker.getTrackerParams().lang = call.argument(VALUE)
            GET_LANG -> MyTracker.getTrackerParams().lang
            SET_CUSTOM_USER_IDS -> MyTracker.getTrackerParams().customUserIds = call.argument<List<String>>(VALUE)?.toTypedArray()
            GET_CUSTOM_USER_IDS -> MyTracker.getTrackerParams().customUserIds?.toList()
            SET_EMAILS -> MyTracker.getTrackerParams().emails = call.argument<List<String>>(VALUE)?.toTypedArray()
            GET_EMAILS -> MyTracker.getTrackerParams().emails?.toList()
            SET_PHONES -> MyTracker.getTrackerParams().phones = call.argument<List<String>>(VALUE)?.toTypedArray()
            GET_PHONES -> MyTracker.getTrackerParams().phones?.toList()

            else -> {
                result.notImplemented()
                return
            }
        }

        result.success(if (value is Unit) null else value)
    }

    private fun init(call: MethodCall) {
        contextRef?.get()?.also { MyTracker.initTracker(call.argument(INIT_PARAM_ID)!!, it as Application) }

        activityRef?.get()?.also { MyTracker.trackLaunchManually(it) }
    }

    companion object {
        const val CHANNEL_NAME = "_mytracker_api_channel"

        const val INIT_METHOD = "init"
        const val INIT_PARAM_ID = "id"

        const val FLUSH_METHOD = "flush"

		const val GET_INSTANCE_ID = "getInstanceId"

        const val TRACK_EVENT_METHOD = "trackEvent"
        const val TRACK_EVENT_NAME = "name"
        const val TRACK_EVENT_PARAMS = "eventParams"

        const val TRACK_LOGIN_METHOD = "trackLoginEvent"
        const val TRACK_USERID = "userId"
		const val TRACK_VKCONNECTID = "vkConnectId"

        const val TRACK_REGISTRATION_METHOD = "trackRegistrationEvent"

        const val IS_DEBUG_MODE_METHOD = "isDebugMode"

        const val SET_DEBUG_MODE_METHOD = "setDebugMode"

        const val GET_ID_METHOD = "getId"

        const val VALUE = "value"

        const val GET_BUFFERING_PERIOD_METHOD = "getBufferingPeriod"

        const val SET_BUFFERING_PERIOD_METHOD = "setBufferingPeriod"

        const val GET_FORCING_PERIOD_METHOD = "getForcingPeriod"

        const val SET_FORCING_PERIOD_METHOD = "setForcingPeriod"

        const val GET_LAUNCH_TIMEOUT_METHOD = "getLaunchTimeout"

        const val SET_LAUNCH_TIMEOUT_METHOD = "setLaunchTimeout"

        const val SET_PROXY_HOST_METHOD = "setProxyHost"

        const val SET_TRACKING_ENVIRONMENT_ENABLED = "setTrackingEnvironmentEnabled"
        const val IS_TRACKING_ENVIRONMENT_ENABLED = "isTrackingEnvironmentEnabled"

        const val SET_TRACKING_LAUNCH_ENABLED = "setTrackingLaunchEnabled"
        const val IS_TRACKING_LAUNCH_ENABLED = "isTrackingLaunchEnabled"

        const val SET_TRACKING_LOCATION_ENABLED = "setTrackingLocationEnabled"
        const val IS_TRACKING_LOCATION_ENABLED = "isTrackingLocationEnabled"

        const val GET_AGE = "getAge"
        const val SET_AGE = "setAge"

        const val GET_GENDER = "getGender"
        const val SET_GENDER = "setGender"

        const val GET_LANG = "getLang"
        const val SET_LANG = "setLang"

        const val GET_CUSTOM_USER_IDS = "getCustomUserIds"
        const val SET_CUSTOM_USER_IDS = "setCustomUserIds"

        const val GET_EMAILS = "getEmails"
        const val SET_EMAILS = "setEmails"

        const val GET_PHONES = "getPhones"
        const val SET_PHONES = "setPhones"
    }
}
